### PR TITLE
Add some more test coverage

### PR DESCRIPTION
**Pull request by timj**
* Test astResample (use smaller variant of testhuge)
* Add some more example FITS headers

### CODE REVIEW

File: src/fitschan.c [hunk 1]
```diff
@@ -1286,6 +1286,24 @@ f     - AST_WRITEFITS: Write all cards out to the sink function
...
+*        The original code used a VELO-%3c wildcard match followed by
+*        CardName to determine which VELO-xxx keyword was found, but
+*        CardName returns the current card (which has moved on), not
+*        the matched card. Replaced with explicit lookups for each
+*        VELO-xxx variant.
+*        Add missing Ecliptic case in SkyToFits system mapping so
+*        ecliptic coordinates write ELON/ELAT instead of UNLN/UNLT.
+*        Fix memory leak in CLASSFromStore: freqfrm was never annulled
+*        after being used to create a frequency-to-velocity conversion.
+*     23-APR-2026 (TIMJ):
```
Comment (timj) [inline src/fitschan.c:1303]: @dsberry this fundamentally changes how the sky offset headers are read so I would like some comment from you as to whether the dead code was deliberate and I've just broken things, or it was meant to be triggered and I just fixed something.

**Comment by embray** [comment]
Ah, I was about to look into improving the coverage on ast_tester/ itself, especially after my latest PR update complained about a tiny coverage decrease in some of the test files themselves.

Having coverage on the tests is good because it helps to ensure that the tests are actually being run as expected.  But tests in particular result in a lot of false positives for uncovered lines when the tests are passing as intended.  So I was going to take a stab at adding some exclude rules for patterns in the tests to try to make their coverage reports more meaningful.   But that doesn't seem to be at odds with what you're doing here.

**Comment by timj** [comment]
Yes. I was pondering the test coverage because I am making big improvements in the fitschan coverage but it's barely moving the needle because of all the error checks in the new test code.

**Comment by timj** [comment]
This is the current status. I'm starting to lose the will to live but on the other hand edge cases are the most important part of fitschan.c

# fitschan.c Coverage Summary

## Current Status (2026-04-24)
- **Lines**: 90.5% (10549/11653) — up from 78.0%
- **Branches**: 68.4% (7172/10479) — up from 58.1%
- **Tests**: 288 total (139 original + 149 new)
- **Bugs fixed**: 8
  1. FreePV leak in wcsmap.c (astGetNin→mapping.nin in destructor)
  2. TidyOffsets dead code in fitschan.c (IsASkyFrame→astIsASkyFrame)
  3. RESTFREQ GHz/MHz comment detection (re-find card before CardComm)
  4. CLASSFromStore rest frequency uninitialized for galactic coordinates
  5. SetAttrib AltAxes sscanf pattern (%d→%n for string values)
  6. NearestPix rounding: (int)(x+0.5)→round(x) for platform independence
  7. FitsAxisOrder leak/use-after-free in Copy/Delete
  8. LoadFitsChan FindString count 9→10 for KINT type
- **Latent bugs found**: 1 (RFVAL keywords never written by SpecTrans despite
  being read in WcsCelestial — only reachable via hand-crafted headers)

## Uncovered Lines Breakdown
- **Total uncovered**: 1104 executable lines

### Remaining uncovered by function (verified from gcovr)
| Function | Uncovered | Notes |
|---|---|---|
| SpectralAxes | 54 | Grism write (39 — needs non-linear GrismMap), LOG/grism detection overhead (11), non-linear unit mapping (4 — unreachable) |
| AIPSPPFromStore | 47 | Validation/error paths for projections and missing keywords |
| AIPSFromStore | 45 | Validation/error paths for projections and missing keywords |
| CLASSFromStore | 43 | Validation/error paths, 1D→3D axis expansion |
| SetFits | 42 | Typed keyword storage branches (KINT, COMPLEXF, COMPLEXI overwrite paths, error handling) |
| SIPMapping | 40 | SIP distortion write with >2 axis PermMap handling |
| PCFromStore | 38 | Secondary axis celestial parameters, projection params, error paths |
| ComBlock | 37 | CLASS keyword pattern matching for unusual MAXIS/NAXIS |
| SkySys | 30 | 15 dead code (isoff>0 unreachable, unknown system CTYPE), 15 misc |
| SplitMap | 27 | Nested WcsMap cleanup, degenerate mappings |
| FitsGetCom | 27 | Protected method, no internal callers — unreachable from C tests (24 body + 3 dispatch wrapper) |
| WcsFromStore/output | 23 | SREF/SREFP keywords, error paths for incomplete alternates |
| WcsCelestial | 21 | Error/warning paths (unknown projection, duplicate axes) |
| GetAttrib/SetAttrib | 20 | Rarely-used attribute accessor branches |
| GetItem/GetMaxJM/etc | 19 | Store access edge cases |
| CelestialAxes | 19 | Scattered edge cases in celestial axis detection |
| Match/MatchFront | 19 | 14 dead code (digit widths 4-9 unused), 5 edge cases |
| MakeIntWorld/CDMatrix | 18 | Degenerate axes (WCSAXES > NAXIS), column-scaling fallbacks |
| WcsNatPole | 15 | Degenerate numerical cases (sin(theta0)=-1, indeterminate LATPOLE) |
| ChangePermSplit | 14 | Recursive PermMap decomposition in CmpMaps |
| MakeIndentedComment | 14 | Native Dump formatting (only used during AST serialization) |
| astFitsChanForId_ | 13 | Fortran interface wrapper — unreachable from C tests |
| MakeFitsFrameSet | 12 | Duplicate SpecFrame/SkyFrame detection, unit conversion fallback |
| AddVersion | 11 | SkyRefIs IGNORED path, non-invertable mapping, WorldAxes failure |
| WorldAxes | 11 | Degenerate axis mappings not produced by well-formed FITS |
| DSBSetUp/DateObs | 10 | DSB sideband setup, old date format edge cases |
| SpecTrans | 10 | FITS-PC/IRAF stray keywords, TNX warning, alternate RESTFREQ |
| SIPIntWorld | 9 | SIP distortion inverse write path |
| IRAFFromStore | 8 | TPN rejection, SIN/NCP-equator edge, LONPOLE rejection |
| PutFits | 7 | All defensive dead code: Split pre-validates values with same sscanf |
| NewCard | 6 | COMPLEXF with AST__BAD value, unknown card type |
| OtherAxes | 6 | Column sort for >2 non-celestial axes, singular matrix |
| MakeIntoComment | 6 | Native Dump comment conversion |
| FsetFromStore | 6 | Error handling in store building |
| CnvType | 5 | Error branches unreachable via public API |
| FindBasisVectors | 4 | Untriggerable from valid FITS (needs undefined IWC at grid centre) |
| WcsSkyFrame | 4 | FK4-no-equinox default, unknown RADESYS error |
| NonLinSpecWcs | 4 | Unknown algorithm code, unknown spectral system (internal errors) |
| IsMapLinear | 4 | MapSplit failure fallback |
| CardData/CardFlags | 3 | Null data/flags edge cases |
| Dump | 2 | Unknown encoding (internal error), FitsTables |
| Load | 2 | First-time vtab init (one-shot) |
| GetEncoding | 2 | Encoding auto-detection edge case |

### Classification of remaining uncovered lines

**Dead code / unreachable (~97 lines):**
- FitsGetCom (27 lines) — protected virtual method with no internal callers; only reachable from external apps using AST internals
- Match digit-widths 4-9 (14 lines) — no keyword pattern uses these
- SkySys isoff>0 block (15 lines) — offset SkyFrames excluded by IsASkyFrame domain check
- PutFits value parse errors (7 lines) — Split pre-validates with same sscanf
- Internal astError paths (~20 lines) — indicate bugs in AST, not user-triggerable
- astFitsChanForId_ (13 lines) — Fortran wrapper, not callable from C

**Validation/error paths (~400 lines):**
- *FromStore functions (AIPSFromStore, AIPSPPFromStore, CLASSFromStore, IRAFFromStore, PCFromStore): unsupported projections, missing keywords, invalid configurations
- WcsCelestial/WcsNatPole/WorldAxes: degenerate numerical cases, duplicate axes
- AddVersion/FsetFromStore: non-invertable mappings, incomplete alternates

**Specialized write paths (~250 lines):**
- SpectralAxes grism (39 lines) — needs non-linear GrismMap over pixel range
- SIPMapping (40 lines) — SIP distortion with >2 axes
- SetFits (42 lines) — complex type overwrite, KINT storage
- PCFromStore secondary celestial (38 lines) — FITS-PC alternate projections
- SplitMap (27 lines) — nested WcsMap decomposition
- ComBlock (37 lines) — CLASS with unusual MAXIS/NAXIS

**Utility / Dump / formatting (~80 lines):**
- MakeIndentedComment, MakeIntoComment, Dump/Load edge cases
- MakeIntWorld degenerate axis handling

### Investigation notes

**PutFits (7 lines):** All uncovered lines are `else` branches after sscanf
parsing of keyword values. The `Split` function pre-validates values using the
same sscanf patterns before PutFits sees them, so these error paths are
unreachable through the public API.

**SkySys isoff>0 (15 lines):** The `isoff > 0` block in SkySys (OFLN/OFLT
CTYPE and offset CTYPE comments) is unreachable because offset SkyFrames have
domain "SKY_ORIGIN"/"SKY_POLE", but SkySys requires domain "SKY" (via
IsASkyFrame check at entry). The OFLN/OFLT output for offset SkyFrames
actually comes from OtherAxes using the SkyFrame's axis symbols.

**SpectralAxes grism write (39 lines):** Requires ExtractGrismMap to find a
GrismMap in the simplified pixel→WCS mapping that produces non-linear output
exceeding IsMapLinear's 1e-7 tolerance. Realistic grism parameters linearise
over typical detector sizes. 1D WAVE-GRI headers fail to read at large
NAXIS1, and direct GrismMap construction fails because MakeFitsFrameSet needs
celestial reference position for FITS-WCS paper III.

**FitsGetCom (27 lines):** Protected virtual method (`astFitsGetCom`) that
retrieves keyword comments. Not exposed in the public `ast.h` header and has
no internal callers within fitschan.c or any other AST source file. Only
reachable by external applications that include fitschan.h directly and call
through the protected interface. The 3-line dispatch wrapper at line 45211 is
also uncovered.

## Source fixes
- `src/wcsmap.c` — FreePV leak fix (astGetNin→mapping.nin in destructor)
- `src/fitschan.c` — TidyOffsets bug fix (IsASkyFrame→astIsASkyFrame)
- `src/fitschan.c` — RESTFREQ GHz/MHz comment fix (re-find card before CardComm)
- `src/fitschan.c` — CLASSFromStore rest frequency fix (move rf check outside
  RADESYS conditional to avoid uninitialized variable for galactic coords)
- `src/fitschan.c` — SetAttrib AltAxes fix (change sscanf from `%d` to
  `%n%*[^\n]%n` pattern — the `%d` stored the parsed integer in `ival` which
  was then reused as a string offset, making string values unreachable)
- `src/fitschan.c` — NearestPix rounding fix (change `(int)(x+0.5)` to
  `round(x)` — the int cast truncates toward zero, giving platform-
  dependent results for pixel values near N.5 due to FP representation
  differences between Linux and macOS)
- `src/fitschan.c` — FitsAxisOrder Copy/Delete fix (deep-copy in Copy
  constructor, free in destructor — was shared pointer / leak)
- `src/fitschan.c` — LoadFitsChan FindString count fix (9→10 to include
  KINT "64 bit integer" type — KINT keywords could be Dumped but not Loaded)

## New test headers (ast_tester/*.head)
ncp, velo-vel, vopt-wav, zopt-frq, cname, skyref, fits-pc-unit, gls,
iraf-longpole, multi-wcs-spec, freq-wav, wavn-frq, mercator, gappt,
atf, equinox-bj, timesys-tdb, zero-cdelt, alt-spectral, helioecl,
fits-pc-full, projp, fk5-noeq, timesys-tt, skyref-alt, axref,
unknown-cel, projp-pv, cd-old, fk4-noeq, unknown-cel2,
tidy-offsets, tidy-pole, date-short, date-frac, equinox-zero,
ssysobs, ener-vel, beta-frq, sip-alt, ukirt-dtai, rfval,
vrad-wav, projp-nopv0, projp-pv0only, restfreq-ghz, pc-sparse,
restfreq-mhz, unknown-spec, crota-3d, axes-swap, cd-cross,
class-noproj, ait-highlatpole, freq-ghz-bary, vopt-geo,
fk4-aips, sfl-origin, vrad-lsrd, class-write, class-gal,
offset-sky-pole, offset-sref, class-arc, class-ait, class-sin,
skyspec3d, skyspec3d-class, offset-sky, azel, vopt-f2w, wave-f2w,
fits-pc-gal, multi-wcs-dup

## testfitschan.c additions
- Dump/Load round-trip with all keyword types: INT, FLOAT, STRING, KINT,
  LOGICAL, COMPLEXF, COMPLEXI, UNDEF; plus Encoding, AltAxes, FitsAxisOrder
  attributes (error 400-409)
- Dump/Load with card flags (USED cards from astRead), UNDEF keyword
  round-trip verification, default Encoding/AltAxes load paths (error 410-416)
- CnvType cross-type conversion tests (error 700-791)
- LOGICAL(false)→STRING edge case (error 790-791)
- All 8 TIMESYS variants loop (error 810-817)
- Ill-conditioned SFL leak test (error 800)
- Non-linear spectral VOPT-F2W, WAVE-F2W, WAVN-W2F, AWAV-F2A round-trips
  (error 606-655)
- Non-linear spectral VELO-F2V, FREQ-A2F, FREQ-V2F (error 660-665)
- Spectral axis Label + ObsGeo write (error 670-673)
- LOG spectral algorithm via MathMap exponential mapping (error 680-684)
- SplitMap linear-sky → CAR write test (error 850-857)
- SplitMap constant-sky → TAN write test (error 860-862)
- FitsAxisOrder write test (error 890-892)
- PutCards bulk card insertion test (error 893)
- DUT1/DTAI read test (error 900-901)
- AddEncodingFrame CLASS VRAD/FK4 tests (error 910-915)
- MakeFitsFrameSet 1D SpecFrame with RefRA/RefDec (error 920-923)
- SkySys axis label write test (error 930-932)
- SkySys helioecliptic write test (error 940-941)
- Old DATE-OBS dd/mm/yy format write test (error 950-951)
- FITS-PC write with alternate axes via 1D spectral FrameSet (error 960-963)
- Offset SkyFrame write with SkyRef/SkyRefIs verification (error 971-973)

## What Would Move the Needle Next

Of the 1104 remaining uncovered lines, approximately:
- **~97 are dead code** or unreachable (protected methods, Fortran wrappers, internal errors)
- **~400 are validation/error paths** in the *FromStore encoding functions
- **~250 are specialized write paths** (grism, SIP >2D, FITS-PC celestial alternates)
- **~80 are utility/Dump/formatting** code

The highest-value remaining targets:
1. **SetFits overwrite paths** (42 lines) — exercising complex type overwrites
   and KINT storage via astPutFits with overwrite=1 on existing cards.
2. **SIPMapping >2D** (40 lines) — needs a 3D+ FrameSet with SIP distortion
   and non-celestial axes. Requires careful Mapping construction.
3. **PCFromStore celestial alternates** (38 lines) — needs a multi-Frame
   FrameSet with celestial axes that share the same PC matrix across versions.
4. **ComBlock CLASS** (37 lines) — unusual MAXIS/NAXIS combinations in CLASS
   headers. Main paths are covered.
5. **SpectralAxes grism write** (39 lines) — see investigation notes.
   Fundamentally hard to trigger from the public API.

**Comment by timj** [comment]
I've just noticed the vast number of test headers I had to add on this PR... I think I will have to reorganize the ast_tester directory at some point.

**Comment by timj** [comment]
There is a status report on #44 but fundamentally I think I've improved the FitsChan coverage as much as is reasonable. There is a lot of dead code paths that Claude couldn't work out how to trigger. It spent a lot of time on the GrismMap case but couldn't work out how to make it hit the non-linear path.

I think this is ready to merge. A few more buffer overruns were fixed.

**Comment by dsberry** [comment]
@timj Feel free to merge when you think it's ready. Probably best just to call me in when you have a specific  question or issue you want my input on.

**Comment by timj** [comment]
There was just the IsASkyFrame vs astIsASkyFrame logic change that changed how offsets worked and unlocked a whole chunk of code.

**Comment by dsberry** [comment]
Yes I had a look at that one. I think it's a fair cop. When I added support for polarisation data (mainly Stokes parameter images) I used a SkyFrame with Domain set to POLANAL. I think I did a global edit of astIsASkyFrame -> IsASkyFrame in order to avoid using these POLANAL Frames as SkyFrames. But I obviously missed the fact that Domain is changed when storing sky offsets. So I think your fix is fine. The fact that this was never noticed suggests no one ever used FITS to store sky offsets.

**Comment by timj** [comment]
Okay. Thanks. I have merged it. I will do a reorganization later. Notes are in #44 so we do not forget the discussion on unreachable code.

The biggest gap in coverage was the non-linear GrismMap which Claude could not work out how to trigger.
